### PR TITLE
CI: Add a new workflow that tries to automatically update nix hashes

### PR DESCRIPTION
**Pull request by nomeata**
Usually, when we upgrade some dependencies, we have to manually update the hashes. This involves
 * changing the hash to something else (e.g. changing one character)
 * running the build locally
 * observing the error message with the actual hash
 * pasting that
 * commit and push.

The `nix-update` tool can automate that.

This PR makes Github do that on all pushes (even to feature branches), and if needed, push a fix to that branch.

With #2761 adding even more fixed output derivations, this may be more relevant.

### CODE REVIEW

**Comment by nomeata** [comment]
This seems to work, I suggest we get it into `master` to see if it works reliably, and then possibly expand to other hashes in the repo, in particular the RTS hashes.

It breaks they hydra build, because it pulls in code that’s not on DFINITY’s whitelist. Maybe time to turn hydra off finally? (See #2771)

**Review by ggreif — COMMENTED** [review]
This needs some prose in the explanation otherwise I don't know what I am reviewing :-)

**Comment by nomeata** [comment]
True; updated the PR description

**Review by ggreif — APPROVED** [review]
Great addition! Thanks!

**Comment by ggreif** [comment]
@nomeata will we merge this?

**Comment by nomeata** [comment]
No hurry, maybe better when I'm back from traveling.

**Comment by nomeata** [comment]
Looks like I lost my commit rights here too? Is that intentional?

**Comment by ggreif** [comment]
> Looks like I lost my commit rights here too? Is that intentional?

No way. What are the symptoms you see? E.g. unable to resolve the conflict?

File: .github/workflows/update-hash.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  update-hash:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        # This is needed to be able to push and trigger CI with that push
+        token: ${{ secrets.NIV_UPDATER_TOKEN }}
+    - uses: cachix/install-nix-action@v13
+      with:
+        nix_path: nixpkgs=channel:nixos-21.05
```
Comment (ggreif) [inline .github/workflows/update-hash.yml:16]:
```suggestion
        nix_path: nixpkgs=channel:nixos-21.11
```

File: nix/default.nix [hunk 2]
```diff
@@ -43,6 +43,11 @@ let
           }
         )
 
+        ( self: super: {
```
Comment (ggreif) [inline nix/default.nix:46]: this is not needed any more: https://search.nixos.org/packages?channel=21.11&from=0&size=50&sort=relevance&type=packages&query=nix-update

File: nix/nix-update.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
+{ lib
```
Comment (ggreif) [inline nix/nix-update.nix:1]: ditto: `nixpkgs-21.11` has this

**Comment by nomeata** [comment]
Yes, GitHub says a committer has to do this. Also, same with ic-hs: https://github.com/dfinity/ic-hs/pull/65#issuecomment-988250510

**Comment by nomeata** [comment]
I can't commit your suggestions, please apply yourself

**Comment by rossberg** [comment]
@nomeata, I just added you to the repo. Maybe it just happened to work earlier because you were member of the Dfinity core team?

**Comment by nomeata** [comment]
No, I lost that membership months ago, and were manually added to this repo and ic-hs. I assume it was some over-eager security team person or script? Thanks for adding me back.

**Comment by ggreif** [comment]
A jira is open to add back the commit bit.

**Comment by nomeata** [comment]
> A JIRA is open to add back the commit bit.

Andreas already took care of it for motoko, so ic-hs is still missing

File: nix/default.nix [hunk 4]
```diff
@@ -41,7 +41,7 @@ let
               };
             };
           }
-        )
+        ))
```
Comment (ggreif) [inline nix/default.nix:44]:
```suggestion
        )
```

**Comment by ggreif** [comment]
@nomeata the action sez: `error: unrecognised flag '-c'`

File: nix/drun.nix [hunk 5]
```diff
@@ -14,7 +14,7 @@ pkgs:
       # To automate this, try running the following in the nix/ directory
       # nix run -f https://github.com/Mic92/nix-update/archive/master.tar.gz -c nix-update --version=skip drun
```
Comment (ggreif) [inline nix/drun.nix:15]: This comment should be rewritten to refer to the newly-added action.

File: .github/workflows/update-hash.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      with:
+        nix_path: nixpkgs=channel:nixos-21.11
+    - uses: cachix/cachix-action@v10
+      with:
+        name: ic-hs-test
+        authToken: '${{ secrets.CACHIX_AUTH_TOKEN }}'
+    - name: Update drun cargo hash
+      run: |
+        cd nix
+        nix run -f . nix-update -c nix-update --version=skip drun
```
Comment (ggreif) [inline .github/workflows/update-hash.yml:24]:
```suggestion
        nix shell -f . nix-update -c nix-update --version=skip drun
```

File: .github/workflows/update-hash.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      with:
+        nix_path: nixpkgs=channel:nixos-21.11
+    - uses: cachix/cachix-action@v10
+      with:
+        name: ic-hs-test
+        authToken: '${{ secrets.CACHIX_AUTH_TOKEN }}'
+    - name: Update drun cargo hash
+      run: |
+        cd nix
+        nix shell -f . nix-update -c nix-update --version=skip drun
```
Comment (ggreif) [inline .github/workflows/update-hash.yml:24]:
```suggestion
        nix --extra-experimental-features nix-command shell -f . nix-update -c nix-update --version=skip drun
```

File: .github/workflows/update-hash.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+jobs:
+  update-hash:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        # This is needed to be able to push and trigger CI with that push
+        token: ${{ secrets.NIV_UPDATER_TOKEN }}
+    - uses: cachix/install-nix-action@v13
```
Comment (ggreif) [inline .github/workflows/update-hash.yml:14]:
```suggestion
    - uses: cachix/install-nix-action@v15
```

File: .github/workflows/update-hash.yml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    - uses: cachix/cachix-action@v10
+      with:
+        name: ic-hs-test
+        authToken: '${{ secrets.CACHIX_AUTH_TOKEN }}'
+    - name: Update drun cargo hash
+      run: |
+        cd nix
+        nix --extra-experimental-features nix-command shell -f . nix-update -c nix-update --version=skip drun
+    - name: Commit changes
+      uses: EndBug/add-and-commit@v7.4.0
```
Comment (ggreif) [inline .github/workflows/update-hash.yml:26]:
```suggestion
      uses: EndBug/add-and-commit@v7.5.0
```

**Review by ggreif — APPROVED** [review]
This looks like ready for the heavy-duty!

**Comment by ggreif** [comment]
@nomeata you have both bits back :-) Sorry for the grievances!

> Ich bitte um Verzeihung fuer die Umstaende und um das
> Weiterleiten dieser Entschuldigung an Joachim und Andreas.

**Comment by nomeata** [comment]
Thanks for quick fixing :-)

**Pull request merged**